### PR TITLE
adding the ability to select different qualities for png, jpeg and we…

### PR DESCRIPTION
**Pull request by FailedSitcom**
…bp images

## Description

This gives the ability to set different qualities for different image formats:

```
query Example {
    resource {
      coverImage {
        fluid(maxWidth: 100, webp_quality: 90, jpeg_quality: 50, png_quality: 60) {
          ...GatsbyImageSharpFluid_withWebp
        }
      }
    }
  }
```

If none of the format specific qualities are passed, it will fallback to the `quality` argument if passed, otherwise the default quality will be used.

Before this is merged I would also like to add documentation explaining how it works, but as this is my first experience in contributing to Gatsby, I wanted to see if I've gone about this in the right way and if there are any concerns that I'm missing.

## Related Issues

Fixes https://github.com/gatsbyjs/gatsby/issues/10203

### CODE REVIEW

**Comment by coreyward** [comment]
At a glance this looks pretty good, but I think it's important to also update the official Contentful-source plugin that integrates with Gatsby Image with these changes as well. 

https://github.com/gatsbyjs/gatsby/blob/2f4c4608df62e867f646881a00a5141860b7f5e0/packages/gatsby-source-contentful/src/extend-node-type.js#L62-L79

File: packages/gatsby-plugin-sharp/src/process-file.js [hunk 1]
```diff
@@ -211,10 +217,12 @@ const compressPng = (pipeline, outputPath, options) =>
...
-            quality: `${options.quality}-${Math.min(
-              options.quality + 25,
-              100
-            )}`, // e.g. 40-65
+            quality: options.png_quality
+              ? `${options.png_quality}-${Math.min(
+                  options.png_quality + 25,
+                  100
+                )}`
+              : `${options.quality}-${Math.min(options.quality + 25, 100)}`, // e.g. 40-65
```
[the remark this answers is not in the document]
Comment (FailedSitcom) [reply packages/gatsby-plugin-sharp/src/process-file.js]: @coreyward nice call

**Comment by FailedSitcom** [comment]
Thank you for the feedback @coreyward! I saw your mention of the Contentful integration on the original issue, and wanted to look into it before requesting a further review (but am very pleased you were able to point me to the relevant part of the code).

If I implemented it this week, would you be interested in taking a look?

**Comment by coreyward** [comment]
I'd be happy to help review, but I'm not actually that familiar with that part of the codebase. I'm confident that one of the members of the core team will review and provide additional guidance as well—they do a great job with new contributors.

**Comment by FailedSitcom** [comment]
Well I appreciate the time you've already taken to look over it, and will definitely check-in with the core team when I've addressed the points you raised.

Thanks again!

File: packages/gatsby-plugin-sharp/README.md [hunk 2]
```diff
@@ -67,6 +70,9 @@ Automatically create sizes for different resolutions — we do 1x, 1.5x, and 2x.
 - `width` (int, default: 400)
 - `height` (int)
 - `quality` (int, default: 50)
+- `jpeg_quality` (int)
+- `png_quality` (int)
+- `webp_quality` (int)
```
Comment (vladar) [inline packages/gatsby-plugin-sharp/README.md]: Maybe replace underscores with camelCase as this looks inconsistent now (i.e. `png_quality` but `pngCompressionLevel`)?

Comment (vladar) [reply packages/gatsby-plugin-sharp/README.md]: We also need a note in docs about this feature. I guess the original issue description is a good start for it.

File: packages/gatsby-plugin-sharp/src/plugin-options.js [hunk 3]
```diff
@@ -11,6 +11,9 @@ const pluginDefaults = {
 
 const generalArgs = {
   quality: 50,
+  jpeg_quality: ``,
+  png_quality: ``,
+  webp_quality: ``,
```
Comment (vladar) [inline packages/gatsby-plugin-sharp/src/plugin-options.js]: Having those args as empty strings seem a bit weird. Maybe we could change them to `null` (or `undefined`) and make sure they are handled properly in the `healOptions` function (with `options.quality` as a fallback as suggested by @coreyward below).

**Comment by vladar** [comment]
@FailedSitcom thanks for the PR - it will be beneficial for the community once merged.

Please let me know if I can help you move this forward?

This could be a good case for a [pair programming](https://www.gatsbyjs.org/contributing/pair-programming/) session. Or, if you don't have time, you could provide us push-access to your branch so that we could do those tweaks and merge?

**Comment by FailedSitcom** [comment]
@vladar thank you for your helpful suggestions and encouragement. I'd sadly been too busy over the last couple of weeks to implement the changes you suggested, but would be more than happy to this weekend.

If something comes up and I'm unable to do so, I'd be more than happy to provide push-access.

Thank you (and the awesome Gatsby team) again!

**Comment by vladar** [comment]
Sounds good! Thanks for doing this!

**Pull request merged**